design/node-agent-concurrency.md

```diff
@@ -0,0 +1,78 @@
+
+### Per-node concurrent number
+We allow users to specify different concurrent number per node, for example, users can allow 3 concurrent instances in Node-1, 2 instances in Node-2 and 1 instance in Node-3. This is for below considerations:
+- The resources may be different among nodes. Then users could specify smaller concurrent number for nodes with less resources while larger number for the ones with more resources
+- Help users to isolate critical environments. Users may run some critical workloads in some specified nodes, since VGDP instances may take large resource consumption, users may only want to run them in the other nodes or run less number of instances in the nodes with critical workloads
+
+The range of Per-node concurrent number is the same with Global concurrent number.  
+Per-node concurrent number is preferable to Global concurrent number, so it will overwrite the Global concurrent number for that node.  
+Per-node concurrent number is implemented in a separate configMap named ```node-agent-configs```. This configMap is not created by Velero, users should create it manually on demand. The configMap should be created before node-agent server starts, otherwise, node-agent server should be restarted in order to make the effect.  
+
+The ```node-agent-configs``` configMap may be used for other purpose of configuring node-agent in future, at present, there is only one kind of data ```data-path-concurrency```. Its data struct is like below: 
+
+```go
+type DataPathConcurrency struct {
+	// ConfigRules specifies the configs by node identified by label selector
+	ConfigRules map[string]int `json:"configRules"`
+}
+```
+
+The key of map of ```ConfigRules``` is a label selector, this means, the nodes are identified by labels.  
+For example, the map could have below elements:
+```
+"kubernetes.io/hostname=node1": 3
+"beta.kubernetes.io/instance-type=Standard_B4ms": 5
+```
+The first element means the node with host name ```node1``` get the Per-node concurrent number of 3.  
+The second element means all the nodes with label ```beta.kubernetes.io/instance-type``` and value ```Standard_B4ms``` get the Per-node concurrent number of 5. 
+At least one node is expected to have a label with the specified key. If no node is with this label, the Per-node configuration makes no effect.  
+If one node falls into more than one elements, e.g., if ```node1``` also has the label ```beta.kubernetes.io/instance-type=Standard_B4ms```, the smallest number (3) will be used.  
+
+### Global data path manager
+As for the code implementation, data path manager is to maintain the total number of the running VGDP instances and ensure the limit is not excceeded. At present, there is one data path manager instance per controller, as a result, the concurrent numbers are calculated separately for each controller. This doesn't help to limit the concurrency among different requesters.  
```

Comment:
Can you help me understand. 

This is saying that the node agent controller will be responsible for the number of concurrent DataUploads/DataDownloads it can handle and ensure this is true? Can you help me understand why this is not the thread of the workers for the controller and needs something in a global manager?

Reply:
> This is saying that the node agent controller will be responsible for the number of concurrent
No, this is the current behavior. We need to change the control to the node-agent server wide.

Reply:
I don't understand. Can you please help me?

Is this because we have two things (DataUploads and DataDownloads) with workers, and you want to split them? 

If this is the case, then isn't there a concern that a node agent will take ownership, because its worker is free just spinning because there are no available "dataManager" threads? Then other node agents that are in the system, that may have free "dataManager" threads, will be sitting idle?

Did I miss us handling this in the design?

Reply:
I might have mislead you. The thing is, in the existing code, there are one dataPathMgr in one controller, so in order to count the overall VGDP instances no matter they are from a DataUpload, DataDownload, PodVolumeBackup or PodVolumeRestore, we need to make a singleton dataPathMgr. See [the existing code](https://github.com/vmware-tanzu/velero/blob/main/pkg/datapath/manager.go)

Reply:
Ok, I do think that, because a node takes ownwership of something and then can determine if it can do work on it, we may want to consider how to make sure this is something that can more adequately be spread out.